In-place edition for project basic information

saraycp opened the pull request:
This is our suggestion:
- in-place edition for the **project**'s basic information
- instead of editing each form element, we propose to edit the whole form (including Title, URL and Description)
- we use Rails UJS
- we add some JavaScript animations for the transitions from content to form and the other way around

It looks like:

![in-place-edition](https://user-images.githubusercontent.com/2581944/76898035-29449e00-6895-11ea-8d12-fc769bcd220b.gif)

Validations are going to be shown in the form as well:

![in-place-edition-validation](https://user-images.githubusercontent.com/2581944/76898119-614be100-6895-11ea-9317-5d53332b0ec8.gif)

Something similar will be done for packages, after we have a quite stable version for projects.

Please take into account that we tried to add little commits to make it easier for you to follow our reasoning. They will be squashed before becoming a real PR.

hennevogel commented on src/api/app/views/webui/project/_basic_info.html.haml:
```diff
@@ -0,0 +1,29 @@
+- if flipper_responsive?
+  .basic-info.mb-3{ class: 'in-place-edition' }
```
The word is `editing`, edition is something else.

saraycp replied:
Oh, good to know! I replace it in the code and as soon as we squash the commits I'll correct the commit's texts as well.

hennevogel commented on src/api/app/views/webui/project/_edit.html.haml:
```diff
@@ -14,3 +14,10 @@
+:javascript
+  $(document).ready(function() {
+    $('.cancel').on('click', function(){
+      location.reload();
```
Hm, I don't understand. Why?

saraycp replied:
This was a first attempt, but in the next commit this is replaced by a normal `remote: true` link to use UJS.

hennevogel commented on src/api/app/assets/javascripts/webui/responsive_ux.js:
```diff
@@ -47,3 +47,15 @@ function toggleTabs(tabLinkContainerId) { // jshint ignore:line
+    $(this).toggleClass('is-invalid');
+    $(this).siblings('.invalid-feedback').remove();
+  });
+}
```
Hm, isn't this something that happens automatically with UJS?

saraycp replied:
The validations weren't shown, so I had to handle it manually. But let me research a bit more...

hennevogel commented on src/api/app/views/webui/project/edit.js.haml:
```diff
@@ -0,0 +1 @@
+$('.in-place-edition').html("#{escape_javascript(render(partial: 'webui/project/edit', locals: { project: @project }))}");
```
Why do you use the `.haml` pre-processor? There is not haml in here.

saraycp replied:
True! There's no reason to use `.haml` to only render a partial. I'll use `.erb` instead.
For both edit and show.

hennevogel commented on src/api/app/controllers/webui/project_controller.rb:
```diff
@@ -226,7 +226,7 @@ def restore
 
   def edit
-    authorize @project, :update?
+    authorize @project, :edit?
```
Why? Do you expect this this to be different?

BTW you can also skip the method argument and it'll be derived from the controller method name you call this in.

saraycp replied:
I just created `:edit?` to be consistent with names, but it's true it's exactly the same as using `:update?`. I'll remove the one I created.

hennevogel commented on src/api/app/views/webui/project/edit.js.haml:
```diff
@@ -1 +1,7 @@
-$('.in-place-edition').html("#{escape_javascript(render(partial: 'webui/project/edit', locals: { project: @project }))}");
+:plain
```
You would not have to do this if you just skip the `.haml` pre-processor.

saraycp replied:
Ok, `*.js.haml` has been converted to `*.js.erb` and the `:plain`has been removed for both edit and show.

saraycp commented:
@hennevogel I've addressed your feedback, except the one related to [validations](https://github.com/openSUSE/open-build-service/pull/9238#discussion_r394257811), I'm doing more research in that regard.
